### ISSUE TITLE
Strange behavior of cox:nlog-lik

### ISSUE THREAD

**Issue by KOLANICH** [dmlc/xgboost#4005]
When testing XGBoost-powered `lifelines` I saw a strange effect: better  (in the sense of the loss used) hyperparams give worse Harrell concordance.
https://gist.github.com/KOLANICH/99726a2d653168ea77aec9a9b368965f#file-xgboostsurvivalregressionstrangeness-ipynb

the first hyperparams were found with 30 iterations of an optimizer (UniOpt with MSMRM backend), the second ones are found with 10000 iterations of an optimizer (UniOpt with Yabox backend). The second ones have lower (`2.570 > 2.569`) 10-fold crossvalidation mean loss, but much worse concordance (`0.888 > 0.803`).

And in fact this is a common situation: I have tested the stuff on ~ 34 datasets, some of the results have the same issue, this dataset was chosen only as an example.

Is it the correct behavior?

**Comment by slundberg** [comment]
You would hope that the best cox loss also gave a very good C-index, but they are different metrics so I would not be surprised to see some differences. If there was a large loss improvement and/or a large C-index reduction then I would be concerned, but small differences likely just reflect the difference between a rank based ordering score and a cox partial likelihood.

**Comment by KOLANICH** [comment]
>large C-index reduction

Is 8% large enough?

**Comment by slundberg** [comment]
hard to say, that does seem non-trivial. If I were debugging it I would start by comparing to the results from the original gbm R package and see if it does something similar.
